OemPkg/OemConfigPolicyCreatorPei/OemConfigPolicyCreatorPei.c

```diff
@@ -0,0 +1,256 @@
+
+    if (EFI_ERROR (Status)) {
+      DEBUG((DEBUG_ERROR, "%a failed to convert variable entry to var list! - %r\n", __FUNCTION__, Status));
+      ASSERT (FALSE);
+      FreePool (*ConfPolicy);
+      return EFI_ABORTED;
+    }
+
+    Offset += VarListSize;
+  }
+
+  DUMP_HEX(DEBUG_ERROR, 0, *ConfPolicy, Offset, "OSDDEBUG: ");
+
+  if (Offset != NeededSize) {
+    // oops we messed up the math, may have corrupted memory...
+    DEBUG((DEBUG_ERROR, "%a expected ConfPolicy size %x does not match actual size %x!\n", __FUNCTION__, NeededSize, Offset));
+    ASSERT (Offset == NeededSize);
+    FreePool (*ConfPolicy);
+    return EFI_ABORTED;
+  }
+
+  DEBUG ((DEBUG_ERROR, "%a ConfPolicy: %p *ConfPolicy: %p ConfVarListPtr: %p &ConfListPtr: %p\n", __FUNCTION__, ConfPolicy, *ConfPolicy, ConfListPtr, &ConfListPtr));
+
+  return EFI_SUCCESS;
+}
+
+/**
+  Module entry point that will check configuration data and publish them to policy database.
+
+  @param FileHandle                     The image handle.
+  @param PeiServices                    The PEI services table.
+
+  @retval Status                        From internal routine or boot object, should not fail
+**/
+EFI_STATUS
+EFIAPI
+OemConfigPolicyCreatorPeiEntry (
+  IN EFI_PEI_FILE_HANDLE     FileHandle,
+  IN CONST EFI_PEI_SERVICES  **PeiServices
+  )
+{
+  EFI_STATUS  Status;
+  POLICY_PPI  *PolPpi        = NULL;
+  VOID        *ConfPolicy    = NULL;
+  UINT16       ConfPolicySize = 0;
+
+  DEBUG ((DEBUG_INFO, "%a - Entry.\n", __FUNCTION__));
+
+  // First locate policy ppi.
+  Status = PeiServicesLocatePpi (&gPeiPolicyPpiGuid, 0, NULL, (VOID *)&PolPpi);
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a Failed to locate Policy PPI - %r\n", __FUNCTION__, Status));
+    ASSERT (FALSE);
+    return Status;
+  }
+
+  // Oem can choose to do any Oem specific things to config here such as enforcing static only config or
+  // selecting a configuration profile based on some criteria
+
+  Status = CreateConfPolicy (&ConfPolicy, &ConfPolicySize);
+
+  DEBUG ((DEBUG_ERROR, "%a ConfPolicy: %p Size: %d\n", __FUNCTION__, ConfPolicy, ConfPolicySize));
+
+  if (EFI_ERROR (Status) || (ConfPolicy == NULL) || (ConfPolicySize == 0)) {
+    DEBUG ((DEBUG_ERROR, "%a CreateConfPolicy failed! Status (%r)\n", __FUNCTION__, Status));
+    ASSERT (FALSE);
+    return Status;
+  }
+
+  DEBUG ((DEBUG_ERROR, "OSDDEBUG 1\n"));
+
+  // Publish immutable config policy
+  // Policy Service will publish the  gOemConfigPolicyPpiGuid so that the Silicon Policy Creator can consume our
+  // Config Policy and map it to Silicon Policies
+  Status = PolPpi->SetPolicy (&gOemConfigPolicyGuid, POLICY_ATTRIBUTE_FINALIZED, ConfPolicy, ConfPolicySize);
+
+  DEBUG ((DEBUG_ERROR, "OSDDEBUG 2\n"));
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a Failed to set config policy! Status (%r)\n", __FUNCTION__, Status));
+    ASSERT (FALSE);

```

Comment:
In the case that configpolicy returns an error, do you expect the system to halt, or do you want it to continue and the system to continue booting?

Reply:
Per the above comment, I think we can let the system continue and let consumers of config decide what they need to do. If a module decides its config is important enough to halt the system, it has the flexibility to do so if it doesn't find the config policy.

For our model, there will be silicon defaults taken if the config policy isn't found, so this is a legitimate case to continue booting.